### PR TITLE
ctfe: limit hashing of big const allocations when interning

### PR DESCRIPTION
**Pull request by lqd**
Const allocations are only hashed for interning. However, they can be large, making the hashing expensive especially since it uses `FxHash`: it's better suited to short keys, not potentially big buffers like the actual bytes of allocation and the associated 1/8th sized `InitMask`. 

We can partially hash these fields when they're large, hashing the length, and head and tail of these buffers, to
limit possible collisions while avoiding most of the hashing work.

r? @ghost

### CODE REVIEW

**Comment by lqd** [comment]
Cycles and wall-times are interesting here, and show no regression.

Unexpectedly, the `coercions` debug incr-full benchmark was instead a tiny win locally for me. Debuginfo is still a problem to reproduce the profile using the CI artifacts, but after stripping, cg can run both (I'd expect that not to matter for program totals ?) and is still a tiny -0.4% change. 

(I also wonder if there's some weird cg variance on my machine, running this benchmark with the same toolchain, in debug or incremental many times: I had a big +/- 50M instructions change 😱 a few times. I went back to valgrind 0.19 instead of master just in case, but it's worrying)

I'll try only hashing the buffer's length and head, just to see how it looks on the perfbot.

**Comment by lqd** [comment]
Here as well, slightly better on `coercions`, so let's see.

@bors try @rust-timer queue

**Comment by lqd** [comment]
Alright. 

Do you think this could be interesting @oli-obk ?

**Comment by lqd** [comment]
So maybe it was indeed noisy. I've reverted everything back to the initial version of this PR then.
In any case, I'll let you decide what we do with this: r? @oli-obk :)

File: compiler/rustc_middle/src/mir/interpret/allocation.rs [hunk 1]
```diff
@@ -49,6 +50,38 @@ pub struct Allocation<Tag = AllocId, Extra = ()> {
...
+            // And its head and tail.
+            self.bytes[..MAX_BYTES_TO_HASH].hash(state);
+            self.bytes[byte_count - MAX_BYTES_TO_HASH..].hash(state);
+        } else {
+            self.bytes.hash(state);
+        }
+
+        // Hash the other fields as usual.
+        self.relocations.hash(state);
+        self.init_mask.hash(state);
```
Comment (oli-obk) [inline compiler/rustc_middle/src/mir/interpret/allocation.rs:78]: Should probably do something similar for the init mask in the future, right?

File: compiler/rustc_middle/src/mir/interpret/allocation.rs [hunk 2]
```diff
@@ -640,13 +673,42 @@ type Block = u64;
...
     len: Size,
 }
 
+// Const allocations are only hashed for interning. However, they can be large, making the hashing
+// expensive especially since it uses `FxHash`: it's better suited to short keys, not potentially
+// big buffers like the allocation's init mask. We can partially hash some fields when they're
+// large.
+impl hash::Hash for InitMask {
+    fn hash<H: hash::Hasher>(&self, state: &mut H) {
+        const MAX_BYTES_TO_HASH: usize = 64;
```
Comment (oli-obk) [inline compiler/rustc_middle/src/mir/interpret/allocation.rs:689]: Share these magic numbers between the two hash impls

**Comment by oli-obk** [comment]
r=me with constants shared

File: compiler/rustc_middle/src/mir/interpret/allocation.rs (hunk 2 above)
Comment (lqd) [reply compiler/rustc_middle/src/mir/interpret/allocation.rs]: Done. Thanks oli.

**Comment by lqd** [comment]
CI is green. I've adjusted the constants and comments.

@bors r=oli-obk

**Pull request merged**